### ISSUE TITLE
Application endpoint discovery

### ISSUE THREAD

**Issue by arne-broering** [margo/specification#40]
How to inform about exposed app endpoints?
How do I query the application?

**Comment by phil-abb** [comment]
@gunjald can you provide more details on what you mean by this? I watched the recording of Wednesday's call and didn't understand what you meant. It wasn't clear to me who needs to know this information and what they would do with it.

For example: 
- Are you talking about endpoints an end-user would need to know about after deploying an application so they could do things like access the application's UI or web service? 
- Are you talking about how different applications, having previous knowledge of each other, discover the endpoints when the applications are deployed on the same device/cluster?
- Are you talking about how applications, having NO previous knowledge of each other, figure out how to talk to each other?
- Are you talking about how the different containers in a single application deployment figure out how to communicate with each other?

**Comment by gunjald** [comment]
My comment was mainly for the two points and one of them you captured already while other are also valid but we have not defined use cases for them so far as I understand. The two points would be then,

1.  endpoints an end-user would need to know about after deploying an application so they could do things like access the application's UI or web service
2. endpoints  that client applications outside of margo edge device may need to know about of deployed application. These client applications on industrial devices as an example a sensor which could submit the temperature or pressure reading to edge application on desired endpoint of the edge application instance

In general there would be endpoints for different reasons that one or other clients in industrial environment will use to connect with the edge applications and it could be a FQDN (DNS name of the application) representing the edge application endpoint but how this FQDN is generated/managed and resolved by those clients who needs it? And if there are other means than DNS which could be evaluated for this purpose?

**Comment by phil-abb** [comment]
@gunjald 

The application vendor's end-user documentation should provide information about accessing any UIs and APIs in the deployed application. Of course, how the application exposes these is not so straightforward under Margo.

For Kubernetes, I'm aware of the following ways to expose UIs and APIs:
- By deploying a service manifest with the [type LoadBalancer ](https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types) as part of your helm chart.
- By using an ingress controller and deploying an [ingress manifest](https://kubernetes.io/docs/concepts/services-networking/ingress/) as part of your helm chart.
- By deploying manifest to use the [Gateway API](https://kubernetes.io/docs/concepts/services-networking/gateway/) as part of your helm chart.

For Docker Compose, I'm aware of the following ways to expose UIs and APIs:
- By [exposing ports](https://docs.docker.com/get-started/docker-concepts/running-containers/publishing-ports/) that map to the container's internal ports.


Margo complicates this because the application vendors don't know what else is being deployed onto the device:

1. Exposing ports using the Kubernetes service manifest or via Docker Compose
    - The service manifest or Docker Compose file provided by the application vendor indicates what port should be exposed. This means multiple application vendors could try exposing the same port. 

2. Defining and Ingress manifest
    - This is the approach I've always used on Kubernetes. 
    - The ingress manifest uses annotations to control different aspects of the ingress controls (e.g., TLS termination rules). The problem with this is each [ingress controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/), of which there are several, uses different annotations. This means application vendors either need to know which ingress controller is on the device or deploy one as part of their application which doesn't seem like a good idea.
    - The other issue occurs if multiple applications deploy ingress manifests using the same hostname. Several problems can occur if this happens. As a result, each application namespace needs to use its own hostname. One example of a problem is if you're using TLS. The certificate referenced in the ingress manifest belongs to the application's namespace. I've addressed this previously by using an app-specific prefix added to the machine's hostname (e.g. myapp.edge.company.com) as the hostname in the ingress manifest.
    - Ingress routes have several issues, many of which are being addressed by the [Gateway API](https://kubernetes.io/docs/concepts/services-networking/gateway/)

4. Using the Gateway API 
    - I haven't used this yet but I imagine it has the same issue as the ingress manifest if the HTTPRoute hostnames are not unique.
    - This is a newer API and I'm not sure how widely supported it is right now.

**Comment by gunjald** [comment]
Hi Phil,

You have captured the different points very well and I think that has another dimension where I was focusing on that is for example if a port has been exposed by the application via a load balancer on its IP address and that port may not be for application vendor but for the other devices or external applications in that environment who may want to connect with that application. And those external application should have a way to discover which edge application they want to connect with and what is the end point of the edge application(defined in Margo.yaml).

WOS may not may what port has been defined within Helm chart or compose file that should be exposed for such purposes and similarly the external applications doesnt know where to connect with margo edge application instance.

Hypothetically, if WOS would have been aware of a port that a given edge application (A helm chart in Margo.yaml) wanted to expose, it may have created an entry to a DNS server e.g. "margo-test-app.enterprisexxx.com" pointing to the load balancer IP and port.  So the external applications may have just use the "margo-test-app.enterprisexxx.com" to connect with the margo edge application. I would treat such procedures as discovery for the sake of explanation.

But so far any margo application specific information about the application endpoint seems to be contained within the artefacts like Helm charts or compose files and only known to application vendor. But then finally once they are deployed on Margo compatible devices they still need to be discoverable by other consumer application outside of margo devices in my opinion.

**Comment by phil-abb** [comment]
@gunjald 

I think we can split this into two areas:
1. How does Margo deal with the different ways of exposing UIs/APIs
2. How does Margo deal with automatic service discovery

**Comment by gunjald** [comment]
On the first point,
1. How does Margo deal with the different ways of exposing UIs/APIs

As described here (https://gateway-api.sigs.k8s.io/) "Gateway API is an official Kubernetes project focused on L4 and L7 routing in Kubernetes. This project represents the next generation of Kubernetes Ingress, Load Balancing, and Service Mesh APIs. From the outset, it has been designed to be generic, expressive, and role-oriented."

When using Gateway API (Gateway API is an interface, or set of resources, that model service networking in Kubernetes.) to manage ingress traffic, the Gateway] resource defines a point of access at which traffic can be routed across multiple contexts -- for example, from outside the cluster to inside the cluster ([north/south traffic])

From the above description the "gateway api" is one of the possibility (I suppose HTTP based endpoints) how an application can be accessed but that also depends upon the protocols it supports. For UI exposure it may be mostly the HTTP protocol where the applications based routes can be defined to the ingress controller to route the traffic towards the applications pods/containers.

From Margo perspective, assuming Gateway API to be a possibility it needs to be understood what should be the responsibilities of the device manufacturer in terms of the different components and how the traffic routes that are application provider responsibility be configured on device ingress controller. There could be other such aspects as well like TLS terminations that are also relevant.

**Comment by phil-abb** [comment]
@arne-broering  @gunjald @adamqiu-emerson 

As part of our breakout group, let's start capturing all the potential problems we can see if an application attempts to expose a service or UI from Kubernetes and/or Docker-compose. 

Please respond with any potential issues you can think of for the area below, or add new areas if you see something missing.

Once we have them we can consolidate them and start to figure out how which problems Margo may need to solve, if any.

- Docker-compose:
  -  The app vendor's docker-compose file exposes ports
    - Potential problem: 
      - App vendors can try to expose the same ports
      - App vendor may try to expose a port that is in use by something else that isn't a Margo application
      - App vendor may try to expose a reserved/restricted port
      - Potential conflicts if multiple app vendors use the same network names.

- Kubernetes
   - The app vendor's helm chart is creating an ingress route
     - Potential Problems:
       - App vendors might not know what ingress control is installed
       - App vendors might use attributes to control logic but the attributes used may not be compatible with the ingress control on the device.
       - App vendors might try to use the same hostname or route
       - App vendors may use wild cards in the hostname (e.g., *.example.com)
       - App vendors may define an ingress without a hostname to handle all requests.
       - App vendors may try to deploy their own ingress controller as part of their application.
       - App vendors may try to use specific TLS termination types (e.g. passthrough or re-encrypt) that alter the behavior 
   - The app vendor's helm chart is creating a service with the type "LoadBalancer"
     - Potential Problems:
       - Service of this type uses an internal "NodePort" which needs to have unique ports specified. 
   - The app vendor's helm chart is creating a service with the type "NodePort"
     - Potential Problems:
       - App vendors may try to use the same port    
   - The app vendor's helm chart is creating a service that is using an "externalName"
   - The app vendor's helm chart is exposing their application using the Kubernetes Gateway API
     - Potential Problems:
       - The Gateway API is newer and would need to target Kubernetes distributions that support it. 
       - The GatewayClass is scoped to the cluster, instead of a namespace, so there could be conflicts if an app vendor attempts to create a GatewayClass object in their helm chart.
       - The HTTPRoute object has many of the same issues as the ingress with potential conflicts with the hostname, routes, TLS termination rules, etc.
   - There may be some other approaches used that should be investigated

- General Potential Problems
  - There might be a potential for protocol mismatch/conflict. For example conflicts between UDP and TCP.
  - Potential for naming conflicts if multiple applications are attempting to be deployed to the same namespace

- Native WASM

**Comment by g0zilla** [comment]
Using the same names for resources can be a problem as well. Even if K8s names have to be unique only per namespace. Namespace names as such have to be unique. In docker-compose, you don't have namespaces. Here, clashes can happen very easily:

### Docker compose

* Unique identifiers for
  * Container names
  *  Network names
  * Volume names

Further, host environment variables can inserted with {var_a} into the  docker-compose.yaml, e.g. to set environment variables for the services. Of course, the yamls of multiple apps using the same host environment variable, e.g. $PASSWORD, can collide.

**Comment by arne-broering** [comment]
To fully capture the problems indicated above, we probably have to recap how WOS+WOA install applications on docker-compose or on Kubernetes+helm:

If we install successively apps on a WOA, is the entirety of the "installed apps" always captured in one "desired state" managed by margo components? or is the "desired state" defined per single app? So, in the first case, the inspection of the desired state would allow for inconsistency checks.

Can we assume that the WOA is fully controlled by the WOS, i.e., all application installations are going through the WOS? or are there potentially apps installed that by-pass margo components?

**Comment by phil-abb** [comment]
>To fully capture the problems indicated above, we probably have to recap how WOS+WOA install applications on docker-compose or on Kubernetes+helm:

Based on what is currently in the specification there is a one-to-one match with the [Application Description](https://specification.margo.org/app-interoperability/application-package-definition/#application-description) and [Application Deployment](https://specification.margo.org/margo-api-reference/workload-api/desired-state-api/desired-state/) (desired state). So, whenever you want to deploy or update an application an ApplicationDeployment file gets created indicating that application should be deployed or updated.

Based on what is currently in the specification, when the device pulls the desired state it gets back one, or more, ApplicationDeployment YAML files for all new applications deployed, or existing applications updated since the last check. Currently, the specification indicates a Git repository is used for this but this will most likely change. There are a few [alternative proposals](https://github.com/margo/woa-interfaces-wg/tree/main/git-alternatives/proposals) being discussed as part of the breakout group. 

>Can we assume that the WOA is fully controlled by the WOS, i.e., all application installations are going through the WOS? or are there potentially apps installed that by-pass margo components?

We should clarify this. I always assumed this to be the case but we should clearly state it.

**Comment by ajcraig** [comment]
> Can we assume that the WOA is fully controlled by the WOS, i.e., all application installations are going through the WOS? or are there potentially apps installed that by-pass margo components?

I'd like to strongly suggest that the device, which is registered through the Workload Orchestration Agent to the WOS, is managed via a single source of truth, the WOS it is enrolled onto. We will run into many problems if we allow for multiple "controllers" for a single edge entity.  

Issues I see with allowing multiple controllers for a single device:
1. Resource calculation issues
- This means the device capabilities that was sent during onboarding is the full set the WOS has to manage. If there are apps being sideloaded there could be resulting errors for the current deployed workloads, and future workloads.
2. Security issues 
3. Backup/Restore or immutability of the devices configuration.

**Comment by gunjald** [comment]
As per my study there seems to be no direct method to avoid conflicts related to cluster-wide share objects like network ports, host file system, k8s namespaces etc. While there are are optional capabilities from cluster networking perspective like NetworkPolicy, NodePort partitioning between static and dynamic ports but irrespective if the cluster is shared with independent applications none of the available mechanisms seems to offer a solution to resolve such conflicts. 
Traits like inter service communication among  application components is part of the Helm charts or docker compose files and any conflicts with other applications on same device will appear only once the actual instantiation will take place during the WOA interactions with the kubernetes API server. This will be a very late event to determine such conflicts and possibly search for other devices by WFM (WOS) in the network to which can host the failed application. 

Though we have not explored the path where if WFM or WOA allowed to open such artefacts like helm charts or compose files to determine common object conflicts but even that will be very complex and become tightly coupled with the these package management tools and constantly need to be observed for any future changes that may give rise to similar issues.

Out of band management (i.e., without peeking inside the application helm charts of compose files) of such conflicts resolution could be very difficult while considering the fact that the applications are expected to come from unrelated or independent organizations and it will have a high potentials for such conflicts in shared clusters.   

There are initiatives around partitioning a physical k8s cluster in multiple virtual cluster and one of which is open source in Github vCluster (https://github.com/loft-sh/vcluster) that can provide isolated clusters from common infrastructure. However this seems to be an additional layer on top of the base kubernetes layer and hence it may not offer a solution for the problem of conflicts in clusterwide resources if an application is mapped to one of the vCluster.

Other exploratory path could be to study if a such cluster-wide resources can be catalogued and reserved for an application and applications are allowed and verified in someway for the their validity before deployment. But it is just and idea to explore and gather more opinions from the members to look out for a solution.
